test/widgets/user_test.dart

```diff
@@ -78,5 +83,51 @@ void main() {
       check(await actualUrl(tester, avatarUrl)).isNull();
       debugNetworkImageHttpClientProvider = null;
     });
+
+    testWidgets('shows placeholder when image URL gives error', (WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+      prepareBoringImageHttpClient(success: false);
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+      final store = await testBinding.globalStore.perAccount(eg.selfAccount.id);
+
+      final badUser = eg.user(avatarUrl: 'https://zulip.com/avatarinvalid.png');
+      await store.addUser(badUser);
+
+      await tester.pumpWidget(
+        TestZulipApp(accountId: eg.selfAccount.id,
+          child: AvatarImage(userId: badUser.userId, size: 30)));
+      await tester.pumpAndSettle();
+      check(findPlaceholder).findsOne();
+
+      debugNetworkImageHttpClientProvider = null;
+    });
+
+    testWidgets('shows placeholder when user avatarUrl is null', (WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+      final store = await testBinding.globalStore.perAccount(eg.selfAccount.id);
+
+      final userWithNoUrl = eg.user(avatarUrl: null);
+      await store.addUser(userWithNoUrl);
+
+      await tester.pumpWidget(
+        TestZulipApp(accountId: eg.selfAccount.id,
+          child: AvatarImage(userId: userWithNoUrl.userId, size: 30)));
+      await tester.pumpAndSettle();
+      check(findPlaceholder).findsOne();
+    });
+
+    testWidgets('shows placeholder when user is not found', (WidgetTester tester) async {
```

Comment:
nit: organize the tests in the same order as the code they're testing. This case comes before the other error cases in the implementation, so its test goes before the other error tests.